### ISSUE TITLE
[autograd.Function] torch.compile w/ once_differentiable leads to opaque graph break

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#106893]
### 🐛 Describe the bug

Some internal models use once_differentiable. The error message is very opaque:
```py
import torch
from typing import *
from torch.autograd.function import once_differentiable


class ScaleGradient(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    @once_differentiable
    def backward(ctx, grad):
        return grad * 0.5

x = torch.randn(3, requires_grad=True)
def f(x):
    return ScaleGradient.apply(x)
output = torch.compile(f, backend='eager', fullgraph=True)(x)
```
returns
```
Unsupported: inline in skipfiles: ScaleGradient.backward  | wrapper /raid/rzou/pt/debug-cpu2/torch/aut
ograd/function.py

from user code:
   File "<ipython-input-20-149cb1010897>", line 18, in f
    return ScaleGradient.apply(x)
  File "/raid/rzou/pt/debug-cpu2/torch/_dynamo/variables/misc.py", line 241, in trampoline_autograd_bw
d
    return fn_cls.backward(*args, **kwargs)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

```

cc @ezyang @gchanan @kadeng @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov 

### Versions

main

**Comment by voznesenskym** [comment]
> I really want a mechanism to say "given this API in PyTorch that lives in a skipfiles, ignore the fact that it lives in a skipfiles and inline it"

We can add that - skipfiles should not be consulted for allowlist anyway - the whole system needs a revisit. Do you want an escape hatch for now?

**Comment by xw285cornell** [comment]
Just wondering what's the status of this? We're seeing a lot of graph breaks because of this, so will be great to come up with a resolution.

**Comment by ezyang** [comment]
Richard: There are a lot of internal models that have custom autograd functions, and this is making things slower for internal model enablement. Can we fix this in Dynamo, or should we have an escape hatch? We wanted a way to mark a function as being inlinable, even if it is in skipfiles.

**Comment by zou3519** [comment]
Debugging this a bit further, it looks like the problem is that Dynamo decides to skip ScaleGradient.backward because it thinks its filename is in the skipfiles list. The reason is that the once_differentiable decorator returns a function whose .co_code has a filename that is in the skipfiles list.

So we probably need some way to say, "don't skip this function returned from once_differentiable"

**Comment by zou3519** [comment]
Assuming we fix the skipfiles problem, there is another problem. once_differentiable has different behavior depending on if the gradients have requires_grad=True or not. In theory we don't actually know if the gradients have requires_grad=True or not at the time of the forward pass, but we may actually be able to assume they don't because the rest of torch.compile assumes that...

**Comment by bdhirsh** [comment]
> In theory we don't actually know if the gradients have requires_grad=True or not at the time of the forward pass, but we may actually be able to assume they don't because the rest of torch.compile assumes that...

Can dynamo insert an assertion that the gradients actually don't require grad somewhere in the backward graph? (pre-aotautograd)

Or if that's annoying I guess we can always assert directly in AOTAutograd (and add a pointer to the once_differentiable comment)

**Comment by zou3519** [comment]
> > In theory we don't actually know if the gradients have requires_grad=True or not at the time of the forward pass, but we may actually be able to assume they don't because the rest of torch.compile assumes that...
> 
> Can dynamo insert an assertion that the gradients actually don't require grad somewhere in the backward graph? (pre-aotautograd)
> 
> Or if that's annoying I guess we can always assert directly in AOTAutograd (and add a pointer to the once_differentiable comment)

The way I am handling this is:
- torch.compile doesn't support higher order gradients
- as a result, we can assume that the grad_mode during the backwards pass is False
- if the grad_mode is False, then once_differentiable doesn't check if the gradients require grad, so we are safe

**Comment by voznesenskym** [comment]
@yanboliang any updates here?
